Migrate database times to UTC

Throne3d opened the pull request:
Instead of storing timestamps in the database in the local time (e.g. EDT, or if you're in Britain, BST) without an associated timezone, instead save them in UTC without an associated timezone. Migrate pre-existing columns that reference timestamps.

This feels quite risky. I ran it, it seemed to work fine, no timestamps appeared different. Rolling back also seemed to work fine. Unfortunately there doesn't seem to be a way to get the "Olson timezone" (i.e. `America/New_York` instead of EDT) directly using Ruby, so I had to hardcode a map – `Time.now.zone` returns e.g. EDT or EST depending on what time of the year it is, which would then mean the change is off by an hour during half the year. (Or, in my case, it returns BST or GMT depending on what time of the year it is.) `fetch_zone` maps this backwards (lossily and through half-guesswork, but this only needs to run on the production server and possibly our local databases).

It seems to keep times constant for me, with `Europe/London`. Performing the migration and removing `config.active_record.default_timezone = :local` gives the same times as rolling back / not having performed the migration and having that line.

It's kinda blegh. We may have lost data when the server changed timezone (back in like October) for winter, as there would have been an hour where timestamps overlapped, and this isn't going to fix that. (Then again, if you currently load those timestamps they'll probably display the wrong timezone; it has no way to distinguish them. We could plausibly distinguish them by looking at sequential IDs, but I don't think Rails does.)

Throne3d commented:
It's also not super huge priority so I have a feeling we're going to ignore it for a while. :P

Marri commented on db/migrate/20170414171825_change_timestamps_to_utc.rb:
```diff
@@ -0,0 +1,78 @@
+  end
+
+  def up
+    local_zone = fetch_zone
```
Is this just to calculate what the servers are running?

Throne3d replied:
Yes. Or local servers (development servers), as this migration will affect those. (It's not much of a problem if local times get to be a few hours out of sync, I don't think, but seems preferable to try to not.)

Throne3d commented:
"Local time" references the time the server's OS is running, as distinct from "UTC" or, idk, "client time" (the time it is for the user who sent it); the main thing is that the server is saving all the timezones as Eastern in the database when it seems like it would be better practice to save as UTC.

Throne3d commented:
I've updated this per new timestamp columns, and also refactored it slightly to make it make more sense. Testing it locally:

- On master, I get different times shown on the server if I use Europe/London and Europe/Madrid as my timezone (Europe/Madrid shows the hours back by one, since it assumes they're an hour forward from if they were Europe/London times); this is when opening a new terminal and restarting the server between requests.
- On this branch, if I migrate, the times shown stay the same as the ones on master without migration.
- On this branch, I get the same times shown on the server if I use Europe/London and Europe/Madrid as my timezone. (My timezone as on the server is `GMT-05:00 Eastern Time…`)

I think we should do this, in case Heroku decides that they don't need to keep the timezones of their applications constant. The servers seem to use UTC, so I could in fact set this to skip if the detected timezone is UTC, and then we're performing a no-op when we run the migration on the server, and we're also not *currently* encountering any problems around daylight saving times as we would if it were on a system that changed around that, but if they do decide the servers don't need to run UTC then we may encounter weird timestamps in the future.

Throne3d commented:
Here is the output I get when running it locally:

```
$ rake db:migrate
The PGconn, PGresult, and PGError constants are deprecated, and will be
removed as of version 1.0.

You should use PG::Connection, PG::Result, and PG::Error instead, respectively.

Called from /home/throne3d/.rvm/gems/ruby-2.4.1/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:240:in `load_dependency'
== 20170914191425 ChangeTimestampsToUtc: migrating ============================
-- execute("UPDATE audits SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 3.1545s
-- execute("UPDATE board_authors SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0006s
-- execute("UPDATE board_sections SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0005s
-- execute("UPDATE board_views SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), read_at=((read_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0004s
-- execute("UPDATE boards SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0004s
-- execute("UPDATE character_aliases SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0004s
-- execute("UPDATE character_tags SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0004s
-- execute("UPDATE characters SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0019s
-- execute("UPDATE favorites SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0004s
-- execute("UPDATE flat_posts SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0022s
-- execute("UPDATE galleries SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0014s
-- execute("UPDATE gallery_tags SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0009s
-- execute("UPDATE icons SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.2998s
-- execute("UPDATE messages SET read_at=((read_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0009s
-- execute("UPDATE password_resets SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0004s
-- execute("UPDATE post_tags SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0006s
-- execute("UPDATE post_viewers SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0006s
-- execute("UPDATE post_views SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), read_at=((read_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0008s
-- execute("UPDATE posts SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), edited_at=((edited_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), tagged_at=((tagged_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.2932s
-- execute("UPDATE replies SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 9.3919s
-- execute("UPDATE reply_drafts SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0008s
-- execute("UPDATE report_views SET read_at=((read_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0007s
-- execute("UPDATE tags SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0012s
-- execute("UPDATE templates SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0008s
-- execute("UPDATE users SET created_at=((created_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC'), updated_at=((updated_at AT TIME ZONE 'Europe/Paris') AT TIME ZONE 'UTC');")
   -> 0.0010s
== 20170914191425 ChangeTimestampsToUtc: migrated (13.1594s) ==================
```

```
$ rake db:rollback
The PGconn, PGresult, and PGError constants are deprecated, and will be
removed as of version 1.0.

You should use PG::Connection, PG::Result, and PG::Error instead, respectively.

Called from /home/throne3d/.rvm/gems/ruby-2.4.1/gems/activesupport-4.2.9/lib/active_support/dependencies.rb:240:in `load_dependency'
== 20170914191425 ChangeTimestampsToUtc: reverting ============================
-- execute("UPDATE users SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0010s
-- execute("UPDATE templates SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0005s
-- execute("UPDATE tags SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0006s
-- execute("UPDATE report_views SET read_at=((read_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0005s
-- execute("UPDATE reply_drafts SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0004s
-- execute("UPDATE replies SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 11.2267s
-- execute("UPDATE posts SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), edited_at=((edited_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), tagged_at=((tagged_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.3905s
-- execute("UPDATE post_views SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), read_at=((read_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0009s
-- execute("UPDATE post_viewers SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0007s
-- execute("UPDATE post_tags SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0010s
-- execute("UPDATE password_resets SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0010s
-- execute("UPDATE messages SET read_at=((read_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0009s
-- execute("UPDATE icons SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.2976s
-- execute("UPDATE gallery_tags SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0009s
-- execute("UPDATE galleries SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0014s
-- execute("UPDATE flat_posts SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0035s
-- execute("UPDATE favorites SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0005s
-- execute("UPDATE characters SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0023s
-- execute("UPDATE character_tags SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0006s
-- execute("UPDATE character_aliases SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0006s
-- execute("UPDATE boards SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0008s
-- execute("UPDATE board_views SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), read_at=((read_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0010s
-- execute("UPDATE board_sections SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0010s
-- execute("UPDATE board_authors SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris'), updated_at=((updated_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 0.0007s
-- execute("UPDATE audits SET created_at=((created_at AT TIME ZONE 'UTC') AT TIME ZONE 'Europe/Paris');")
   -> 7.9875s
== 20170914191425 ChangeTimestampsToUtc: reverted (19.9250s) ==================
```

It definitely appears to work.

Throne3d commented:
I tested it on staging and nothing changed on boards#index, boards#show, posts#show, messages#index or messages#show between each of these steps:
- prior to deploy or migration
- post deploy, prior to migration
- post deply and migration
so it looks like it changes exactly nothing visibly, which is what we want.

I'm happy to merge and run the migration on prod.
